Update the variants and tpls for SEACAS package.py

mvlopri opened the pull request:
Adds variants to turn off tests
Add variants for some missing TPL options
Add the variables required to build in ~shared

<!--  
Remember that `spackbot` can help with your PR in multiple ways:
- `@spackbot help` shows all the commands that are currently available
- `@spackbot fix style` tries to push a commit to fix style issues in this PR
- `@spackbot re-run pipeline` runs the pipelines again, if you have write access to the repository 
-->

davi0011 committed:
Update the variants and tpls for SEACAS package.py

Adds variants to turn off tests
Add variants for some missing TPL options
Add the variables required to build in ~shared

mvlopri commented:
@psakievich

Commit:
Merge branch 'develop' into update_seacas_package_py

alalazo commented on var/spack/repos/builtin/packages/seacas/package.py:
```diff
@@ -146,19 +146,23 @@ class Seacas(CMakePackage):
     depends_on("cmake@3.17:", when="@:2023-05-30", type="build")
     depends_on("mpi", when="+mpi")
-
+    depends_on("zlib", when="+zlib")
```
Why `zlib` and not `zlib-api`?

gsjaardema replied:
I was not aware of zlib-api.  It looks like it was created in August of 2023 and addresses the question I had asked a few years ago (https://groups.google.com/g/spack/c/ex-_auuyrVo/m/8ilGGingBAAJ?utm_medium=email&utm_source=footer&pli=1).  I agree zlib-api would be preferred here.

psakievich replied:
```suggestion
    depends_on("zlib-api", when="+zlib")
```

mvlopri replied:
I am testing this change currently before pushing up another commit.

gsjaardema commented on var/spack/repos/builtin/packages/seacas/package.py:
```diff
@@ -233,7 +237,8 @@ def cmake_args(self):
         )
-
+        if "~shared" in self.spec:
+            options.append(self.define("Seacas_EXTRA_LINK_FLAGS", "z;dl"))
```
Ideally these should be coming from the non-shared builds of netCDF and HDF5 and not needed here, but that does not seem to be working so we specify them here also.

psakievich replied:
Seems reasonable for a work around. Do you think something is broken? i.e. Is this an ideally, b/c it works with a build external to spack, or a ideally as in it would be nice?

mvlopri replied:
This is done the same way in the SEACAS build system through a configure script right now because netCDF and HDF5 need it there as well. It does not seem to work without them external to Spack either. Ideally, it would be handled in the build systems of netCDF and HDF5.

gsjaardema replied:
`Ideally` as in I think that the dependency chain from HDF5 (built in spack) is not being carried through to the applications that eventually use HDF5.  we are adding the libz, libdl dependency in seacas, but the dependency is at the hdf5 level.  Not sure where the missing link is, but it would be ideal if the dependency were specified at the level of the library that has the dependency and then is carried through.  I don't know if this is an error in the HDF5 CMake, or if there is something we are doing in our CMake that is losing the library dependency. 

For now, this is probably the best solution.

gsjaardema reviewed (approved):
I agree with the zlib-api suggested change. 

The `pamgen` variant pulling in Trilinos might have some ambiguity since Trilinos contains an embedded version of seacas and I don't think there is a variant to disable Trilinos's version of seacas, so not sure what happens in that build...

gsjaardema commented on var/spack/repos/builtin/packages/seacas/package.py:
```diff
@@ -146,19 +146,23 @@ class Seacas(CMakePackage):
     depends_on("mpi", when="+mpi")
-
+    depends_on("zlib", when="+zlib")
+    depends_on("trilinos", when="+pamgen")
```
We also need to modify the Trilinos package.py since it explicitly disables pamgen build and there is no variant available to enable pamgen.  I'm not sure of what is the best change -- I think adding yet another variant which would then allow this to be `depends_on("trilinos~exodus+pamgen", when="+pamgen")`

Then, the mpi complication comes in where we need to depend on the correct +mpi/~mpi build of Trilinos, so maybe something like:
```
    with when("+pamgen"):
        depends_on("trilinos~exodus+pamgen+mpi", when="+mpi")
        depends_on("trilinos~exodus+pamgen~mpi", when="~mpi")
```

psakievich replied:
We should do that as part of this PR then.  It would be nice to have the changes linked in the git history.

mvlopri replied:
This should also be a part of my next commit but I was testing it locally first.

gsjaardema commented:
I think the attached patches would get the pamgen variant working.  This does change the Trilinos package, so would need to get maintainer of that package to approve...
[0002-SEACAS-If-pamgen-then-enable-pamgen-in-trilinos-buil.patch](https://github.com/spack/spack/files/14640374/0002-SEACAS-If-pamgen-then-enable-pamgen-in-trilinos-buil.patch)
[0001-Trilinos-expose-pamgen-building-via-variant.patch](https://github.com/spack/spack/files/14640375/0001-Trilinos-expose-pamgen-building-via-variant.patch)

gsjaardema reviewed (changes requested):
I'm good with this if we:

- Fix the pamgen variant
- Make the `zlib` -> `zlib-api` change.
- Add Trilinos maintainer(s) as reviewers (or mention, but I do see you did that with @psakievich)

psakievich reviewed (commented):
A few questions, and I would like to get the trilinos required changes in with this PR since it is a single concept.

davi0011 committed:
Add pamgen to Trilinos as a variant to support SEACAS

This adds the ability to turn off and on pamgen as needed
through the variant interface for the Trilinos package.py.
Add changes for seacas package.py to build the appropriate
Trilinos variants.
Add zlib-api as depends_on instead of zlib directly for SEACAS
package.py

Commit:
Merge branch 'develop' into update_seacas_package_py

gsjaardema commented on var/spack/repos/builtin/packages/seacas/package.py:
```diff
@@ -361,6 +367,9 @@ def cmake_args(self):
                 [define("TPL_ENABLE_METIS", False), define("TPL_ENABLE_ParMETIS", False)]
             )
 
+        if "+pamgen" in self.spec:
```
I would prefer a slightly different syntax (suggested by Spack developers on previous PRs):
```
options.append(from_variant(project_name_base + "_ENABLE_Pamgen", "pamgen"))
options.append(from_variant("TPL_ENABLE_Pamgen", "pamgen"))
```
Not sure if both are needed -- does the `Seacas_ENABLE_Pamgen` get set if `TPL_ENABLE_Pamgen` is specified?

Also not sure whether we need a `Pamgen_ROOT` so CMake can find Pamgen, or is that already taken care of elsewhere...

mvlopri replied:
It does seem to find pamgen when it builds Trilinos with +pamgen as the configure errors out if it is not built properly. I think the Seacas_ENABLE_Pamgen is unneeded. Changing and testing currently.

gsjaardema replied:
> It does seem to find pamgen when it builds Trilinos with +pamgen as the configure errors out if it is not built properly. I think the Seacas_ENABLE_Pamgen is unneeded. Changing and testing currently.

It looks like there is an incorrect test in the SEACAS CMakelists.txt files in IOSS -- They are testing for `Seacas_ENABLE_Pamgen` instead of `TPL_ENABLE_Pamgen` which is the form used for other TPL's.  I've fixed this in a commit today, but that won't work for previous releases...   Probably need to leave the `Seacas_ENABLE_Pamgen` define in for now and we can tie it to previous versions on the next release...

mvlopri replied:
For now, I will leave both in but make the refactor then. I did refactor it and it is working still.

gsjaardema commented on var/spack/repos/builtin/packages/seacas/package.py:
```diff
@@ -233,7 +238,8 @@ def cmake_args(self):
         )
-
+        if "~shared" in self.spec:
+            options.append(self.define("Seacas_EXTRA_LINK_FLAGS", "z;dl"))
```
This should also probably be `options.append(self.define(project_name_base + "_EXTRA_LINK_FLAGS", "z;dl"))`

davi0011 committed:
Refactor define enabling Pamgen

Refactor the define for EXTRA_LINK_FLAGS.

Commit:
Merge branch 'develop' into update_seacas_package_py

gsjaardema commented on var/spack/repos/builtin/packages/seacas/package.py:
```diff
@@ -361,6 +367,10 @@ def cmake_args(self):
                 [define("TPL_ENABLE_METIS", False), define("TPL_ENABLE_ParMETIS", False)]
             )
 
+        if "+pamgen" in self.spec:
```
The `if "+pamgen" in self.spec:` should not be required anymore with this change.

mvlopri replied:
I misunderstood and this is fixed.

davi0011 committed:
Remvoe the if guard for +pamgen defines

gsjaardema commented on var/spack/repos/builtin/packages/seacas/package.py:
```diff
@@ -215,7 +220,7 @@ def cmake_args(self):
         options.extend(
             [
-                define(project_name_base + "_ENABLE_TESTS", True),
+                define(project_name_base + "_ENABLE_TESTS", from_variant("tests")),
```
This could possibly be `from_variant(project_name_base + "_ENABLE_TESTS", "tests"),` but ok as is...

alalazo replied:
```suggestion
                from_variant(project_name_base + "_ENABLE_TESTS", "tests"),
```

gsjaardema reviewed (approved):
Looks good.  Thanks for putting up with all the comments.

gsjaardema reviewed (approved):
Accidentally clicked request new review button.  All looks good.

gsjaardema commented:
@alalazo @psakievich  I don't have commit privilege, would one of you be able to review this.
Thanks.

alalazo committed:
Fix call to define_from_variant
